### ISSUE TITLE
iOS Example

### ISSUE THREAD

**Issue by sbryfcz** [facebookarchive/caffe2#298]
This is really just a question/enhancement. I'd love to see an example of what needs to happen to build an XCode/iOS project using this library. I was able to build_ios just fine, but its unclear what the next steps are. Thanks for the great work.

**Comment by Yangqing** [comment]
Yup, would love to create examples as follow-ups in the next weeks; cc'ing @bwasti who's our mobile expert.

**Comment by mjohnst** [comment]
I have started attempting an integration of the C++ API into an Objective-C++ project. I can post more when I've gotten something running inference on a device.

**Comment by BrianOn99** [comment]
@mjohnst Thanks in advance!  Please show your result once you get something working!

**Comment by abjurato** [comment]
@bwasti, another couple of iOS-related questions:
1. are you planning Metal support? If yes, would it be based on MPSCNN?
2. are you planning any Swift wrapper?

**Comment by RobertBiehl** [comment]
@mjohnst Hey, I'm attempting the same thing.
I built the static library with build_ios.sh and am now trying to build a pod on top of that.

My problem is when referencing the header
`#include "caffe2/core/predictor.h"` in an _Objective C/Objective C ++_  file, this leads to the following errors:
<img width="939" alt="screen shot 2017-04-22 at 11 08 05" src="https://cloud.githubusercontent.com/assets/446619/25303078/0307ec8e-274c-11e7-8fc8-6a650b28eefd.png">

I believe this is due to the fact that the required header has a chain dependency to a header `#include "caffe2/core/net.h"` which contains actual implementation and references c++ libs which might collide with what's been used when building the static lib.

**Comment by RobertBiehl** [comment]
@abjurato A Swift wrapper is what I'm trying to build. Just a simple plug and play pod ;)
It seems that I've become soft when it comes to integrating unwieldy c++ libs...

I'll post when I get there...

**Comment by jquave** [comment]
I've been able to get this up and running: [caffe2 on iOS tutorial](http://jamesonquave.com/blog/caffe2-on-ios-deep-learning-tutorial/).

**Comment by RobertBiehl** [comment]
Thanks @jquave.
I hope I'll have simple to use solution soon.

```
2017-04-22 23:30:29.201545+0200 Caffe2Test[4036:2762929] Inference results:
0: birdhouse - 31.4025%
1: carton - 20.0337%
2: crossword puzzle, crossword - 2.15863%
3: perfume, essence - 1.46789%
4: perfume, essence - 1.46789%
```

At least I got my first inference running on the device! 😄

**Comment by jquave** [comment]
@RobertBiehl can you post what you did to get the models working?

**Comment by jquave** [comment]
I've since been able to successfully run squeezenet and identify objects in images on my iPhone 7. Updated the [Caffe2 iOS tutorial](http://jamesonquave.com/blog/caffe2-on-ios-deep-learning-tutorial/) to reflect that.

**Comment by ankghost0912** [comment]
@jquave thanks for the tutorial. I followed your tutorial, and also passed the `-Wl -force_load libCaffe2_CPU.a` flag  to the linker, but it gives me a Mach-O error. Did you encounter this?

**Comment by jquave** [comment]
@ankghost0912 no Mach-O error. What's the error? Also are you building to device? You need to modify platform in order to build a simulator library.

**Comment by ankghost0912** [comment]
Yes I'm building for the device. Did you build for a simulator? First, Xcode gives a `ld warning: Ignoring file path-to-root/libCaffeNNPack.a` and other such libraries and expectedly reports a linker error like: `  "caffe2::FLAGS_caffe2_print_blob_sizes_at_exit", referenced from:`. I get that you didn't encounter anything like that, right?

**Comment by RobertBiehl** [comment]
I'll see if I can share my working Xcode project when I get up. 

It's still a mess, and my goal was to get at least simultaneous simulator+device build working. Currently you'll. Need to fire up build_ios.sh for each when you switch architectures.

**Comment by RobertBiehl** [comment]
@jquave I think you might want to use BGR, not RGB colors:
https://caffe2.ai/docs/tutorial-image-pre-processing.html#null__caffe-uses-bgr-order

You tutorial code suggests you only take out the alpha channel: `cv::cvtColor(src_img, bgra_img, CV_RGBA2RGB);`

Also you need to make sure you have the correct number of channels when you come from an uiimage.
I do it like this:
```objc
UIImage* image = [UIImage imageNamed:@"cat"];
CFDataRef pixelData = CGDataProviderCopyData(CGImageGetDataProvider(image.CGImage));
const UInt8* pixels = CFDataGetBytePtr(pixelData);
...
inputPlanar[i * predWidth + j + 0 * size] = (float) pixels[(_i * width + _j) * 4 + 2]; // B
inputPlanar[i * predWidth + j + 1 * size] = (float) pixels[(_i * width + _j) * 4 + 0]; // R
inputPlanar[i * predWidth + j + 2 * size] = (float) pixels[(_i * width + _j) * 4 + 1]; // G
```

**Comment by jquave** [comment]
@RobertBiehl thanks for that. I thought it was RGB, but this OpenCV line converts to whatever color space and drops the alpha channel like you mention. It's not necessary to do the work in the loop I think:

```
cv::cvtColor(src_img, bgra_img, CV_RGBA2BGR);
```

The loop is probably something that can be skipped entirely, and just resize in OpenCV as well, I jsut had not taken the time to wrangle the types for that to be loadable in to Caffe2. I'm sure it would work better and use less memory

**Comment by RobertBiehl** [comment]
@abjurato and anyone else playing around with caffe2 on iOS: I'm in the process of creating a pod. If you want to try it out, it's here: [Caffe2Kit](https://github.com/RobertBiehl/caffe2-ios).

**Comment by RobertBiehl** [comment]
When I'm trying to build for simulator with `IOS_PLATFORM=SIMULATOR scripts/build_ios.sh` I get the following error:
```c
[  2%] Building C object CMakeFiles/CAFFE2_NNPACK.dir/third_party/NNPACK/src/neon/blas/conv1x1.c.o
/usr/bin/gcc  -DCAFFE2_UNIQUE_LONG_TYPEMETA -DCAFFE2_USE_EIGEN_FOR_BLAS -DEIGEN_MPL2_ONLY -I/Users/Robert/Documents/Workspace/caffe2-swift/lib/caffe2/build_ios_pod_x86_64 -isystem /Users/Robert/Documents/Workspace/caffe2-swift/lib/caffe2/third_party/protobuf/src -isystem /Users/Robert/Documents/Workspace/caffe2-swift/lib/caffe2/third_party/NNPACK/include -I/Users/Robert/Documents/Workspace/caffe2-swift/lib/caffe2/third_party/NNPACK/src -I/Users/Robert/Documents/Workspace/caffe2-swift/lib/caffe2/third_party/NNPACK_deps/FP16/include -I/Users/Robert/Documents/Workspace/caffe2-swift/lib/caffe2/third_party/NNPACK_deps/FXdiv/include -isystem /Users/Robert/Documents/Workspace/caffe2-swift/lib/caffe2/third_party/NNPACK_deps/pthreadpool/include -I/Users/Robert/Documents/Workspace/caffe2-swift/lib/caffe2/third_party/NNPACK_deps/psimd/include -isystem /Users/Robert/Documents/Workspace/caffe2-swift/lib/caffe2/third_party/eigen -isystem /Users/Robert/Documents/Workspace/caffe2-swift/lib/caffe2/third_party/pybind11/include -I/Users/Robert/Documents/Workspace/caffe2-swift/lib/caffe2  -std=c11 -O3 -DNDEBUG -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator10.3.sdk   -o CMakeFiles/CAFFE2_NNPACK.dir/third_party/NNPACK/src/neon/blas/conv1x1.c.o   -c /Users/Robert/Documents/Workspace/caffe2-swift/lib/caffe2/third_party/NNPACK/src/neon/blas/conv1x1.c
In file included from /Users/Robert/Documents/Workspace/caffe2-swift/lib/caffe2/third_party/NNPACK/src/neon/blas/conv1x1.c:5:
In file included from /Users/Robert/Documents/Workspace/caffe2-swift/lib/caffe2/third_party/NNPACK/include/nnpack/arm_neon.h:3:
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../lib/clang/8.1.0/include/arm_neon.h:28:2: error: "NEON support not enabled"
#error "NEON support not enabled"
 ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../lib/clang/8.1.0/include/arm_neon.h:48:24: error: 'neon_vector_type' attribute is not supported for this target
typedef __attribute__((neon_vector_type(8))) int8_t int8x8_t;
                       ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../lib/clang/8.1.0/include/arm_neon.h:49:24: error: 'neon_vector_type' attribute is not supported for this target
typedef __attribute__((neon_vector_type(16))) int8_t int8x16_t;
                       ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../lib/clang/8.1.0/include/arm_neon.h:50:24: error: 'neon_vector_type' attribute is not supported for this target
typedef __attribute__((neon_vector_type(4))) int16_t int16x4_t;
                       ^
...
fatal error: too many errors emitted, stopping now [-ferror-limit=]
20 errors generated.
make[2]: *** [CMakeFiles/CAFFE2_NNPACK.dir/third_party/NNPACK/src/neon/blas/conv1x1.c.o] Error 1
make[1]: *** [CMakeFiles/CAFFE2_NNPACK.dir/all] Error 2
make: *** [all] Error 2

```

Is there any build variable I need to set to disable Neon support in NNPACK?

**Comment by RobertBiehl** [comment]
It looks like caffe sometimes includes NEON headers without checking if NEON is available, see https://github.com/caffe2/caffe2/blob/129d563f204a3f692ee2907696f9c62e99e55df8/caffe2/operators/conv_transpose_op_mobile_impl.h
```c++
// conv_transpose_op_impl.h is the templated implementation of the
// conv_transpose_op.h file.
#ifndef CAFFE2_OPERATORS_CONV_TRANSPOSE_MOBILE_OP_IMPL_H_
#define CAFFE2_OPERATORS_CONV_TRANSPOSE_MOBILE_OP_IMPL_H_

#include "caffe2/core/common.h"

#ifndef CAFFE2_MOBILE
#error "mobile build state not defined"
#endif

#if CAFFE2_MOBILE

#include "caffe2/core/logging.h"
#include "caffe2/operators/conv_op_shared.h"
#include "caffe2/operators/conv_transpose_op_mobile.h"
#include "caffe2/utils/cpu_neon.h"
#include "caffe2/utils/fixed_divisor.h"
#include "caffe2/utils/math.h"
...
```
And then later on it checks for NEON
```c++
#ifdef __ARM_NEON__
    // We vectorize the loop within the row
    {
...
```

**Comment by Yangqing** [comment]
This is really nice tutorial @jquave , thanks so much!

@RobertBiehl - for the NNPACK engine, unfortunately you will need to build it with OS instead of SIMULATOR. The current NNPACK build script automatically assumes arm build and it is not compatible with the simulator (which is x86) yet, and you are seeing the issue that the x86 simulator compiler does not support neon intrinsics. To build with simulator, do `-DUSE_NNPACK=OFF`.

@abjurato - we are actually in the process of publishing our metal / mpscnn code. There are a few syncing and building system issues to be sorted out (we want to make sure oss uses cmake, internally we use buck and some ios extensions, but we want to be nice to the community :) ), but you can take a sneak peek here: https://github.com/caffe2/caffe2/pull/215 . @ajtulloch and I are out in France for a conference this week, and we'll have things merged proper later.

**Comment by Yangqing** [comment]
And @abjurato - regarding swift, we haven't thought about it yet (we don't have an in-house swift expert in Caffe2 team) but contributions are most welcome!

**Comment by Yangqing** [comment]
@ajtulloch reminded me that the MPSCNN code is pushed: https://github.com/caffe2/caffe2/tree/master/caffe2/contrib/mpscnn-fb

There is no CMakeLists.txt yet, so if someone would like to add it that would make a very nice PR.

Also, the Qualcomm SNPE library integration is here: https://github.com/caffe2/caffe2/tree/master/caffe2/contrib/snpe-fb you still won't be able to run with SNPE, because that depends on SNPE's runtime that Qualcomm will release at a later date (and we only provide the interface). But the code should give you a good idea of the interface.

**Comment by abjurato** [comment]
@Yangqing regarding objective-c wrapper around C++ (which basically is all we need for convenient work from Swift): how can it be done in this repository? Will it be separate bunch of Objective-C++ files, similar to `caffe2/python` directory, that will be copied somewhere by `build_ios.sh`?

For example, in @RobertBiehl's repository https://github.com/RobertBiehl/caffe2-ios I see where I can contribute to - `src/Caffe2.mm`. How can it be done here?

**Comment by bwasti** [comment]
@abjurato it might be best to put an Objective-C wrapper in a `caffe2/contrib/objc` folder.  If enough wrappers exist it might make sense to create a `caffe2/wrappers/` folder, but for now the `caffe2/contrib` folder is best.

**Comment by tdp2110** [comment]
What commit are people using to run build_ios.sh? I swear I could build it yesterday, but not today. I've wiped and reset my mac, followed the [instructions](https://caffe2.ai/docs/getting-started.html?platform=ios&configuration=compile), and I still get the build error from [here](https://github.com/caffe2/caffe2/issues/419)

**Comment by BrianOn99** [comment]
@RobertBiehl Could you please give some instruction to run to the demo inside Caffe2Kit?  After pod install it complains `caffe2/core/predictor.h` file not found.

**Comment by RobertBiehl** [comment]
@BrianOn99 Yes, sorry about that. It seems like it doesn't install correctly if the pod is referenced locally. I'll fix that.

**Comment by KleinYuan** [comment]
@RobertBiehl Creating a Pod is a great idea!   👍
@BrianOn99 I also met the issue you mentioned and it's because you did local install, which does not work correctly. Then I can successfully build through `pod install` from git path whereas the example swift code has some optional variables handling complaints (pbbly because it's swift 3 now?).  

I also got a working real time detection demo app here/also allowing you to test with your own model via downloading: [Caffe2-iOS](https://github.com/KleinYuan/Caffe2-iOS)

**Comment by abjurato** [comment]
@RobertBiehl, @jquave now caffe2 should be built with bitcode for iOS. Could you check that out and update your tutorials/docs?

**Comment by tdp2110** [comment]
What's the status of the metal code? Does one need to explicitly invoke it? Does it run out of the box? (For example, AFAIK, in caffe2 we need to explicitly invoke CUDA).

**Comment by mistiansen** [comment]
@RobertBiehl Do you have advice for the "caffe2/core/predictor.h file not found" error in demo?

**Comment by KleinYuan** [comment]
@mistiansen I assume it's because your Caffe2 build failed and thus that file is missing.

**Comment by mistiansen** [comment]
@KleinYuan I see. Yes, I get the following error:
**Building caffe2
Building for SIMULATOR
Installing to ../install
CMake Error: The source directory "~/Desktop/caffe2-ios/lib" does not appear to contain CMakeLists.txt.**
This directory contains the caffe2 directory that pulled from the repo. There is a CMakelists.txt in that caffe2 directory, but if I try to move it up a directory it still doesn't work. Any thoughts?

**Comment by RobertBiehl** [comment]
Hey, building of caffe2 should not be necessary for this pod. I included prebuilt binaries.
The header files need to be found though. Can you set up a new project and install the pod in there? I didnt get around to fix the Example project yet.

Rob

**Comment by mistiansen** [comment]
@RobertBiehl Thanks for the response; I will try. I am a noob to XCode; what would differ in the project? Just the paths in the include statements for the caffe2 header files?

**Comment by KleinYuan** [comment]
@mistiansen check [this](https://github.com/KleinYuan/Caffe2-iOS#caffe2-ios-build-failed)

**Comment by RobertBiehl** [comment]
Make sure to follow the README.md
Let me know if you have any questions after ;)

**Comment by RobertBiehl** [comment]
@mistiansen I'm sorry I mixed this up, because I was replying per email. My answers were targeted towards https://github.com/RobertBiehl/caffe2-ios

**Comment by ayyoitsp** [comment]
I didn't see a response for the question about metal support on iOS?  Is there GPU support now or is that still coming?  Is there anything that needs to be done to enable GPU support, or does the iOS build script take care of that?

**Comment by futurely** [comment]
GPU is supported with OpenGL.
https://github.com/caffe2/caffe2/tree/master/caffe2/contrib/opengl

**Comment by CanyonWind** [comment]
@ayyoitsp Please check the new ios [cmakelist](https://github.com/caffe2/caffe2/blob/master/caffe2/contrib/ios/CMakeLists.txt), it looks like metal option is added. Haven't tried it though.

**Comment by KleinYuan** [comment]
@abjurato it's added recently and just simply pull the latest master branch and add 
```
-DUSE_METAL=ON
```
in the [build_ios.sh](https://github.com/caffe2/caffe2/blob/master/scripts/build_ios.sh#L51) and that's it.

**Comment by KleinYuan** [comment]
Trying to implement MPSCNN on Caffe2-iOS and I can build without any problem with `tags/v0.8.1-157-g5b696c12` and addubf the flag to enable use metal.
However, I'm kinda struggling with converting graph to mpscnn.

I am testing with squeezeNet/googleNet on iPhone 7 plus (iOS 10.3.3 / Xcode 8.3.3)

It looks like convert graph to mpscnn is quite straightfroward by just 

```
caffe2::tryConvertToMPSCNN(_initNet, _predictNet, &_mpscnnPredictNet)
```
More details [here](https://github.com/KleinYuan/Caffe2-iOS/blob/gpu/src/caffe2-ios/caffe2-ios/Caffe2.mm#L151).

But it failed and here are the [errors](https://github.com/caffe2/caffe2/blob/417da38a988f2da7900a64dfa8cdfde635b0e8d6/caffe2/contrib/ios/mpscnn/mpscnn_graph.mm#L324):

```
[E mpscnn_graph.mm:324] Caught exception trying to convert NetDef to MPSCNN: [enforce fail at mpscnn_graph.mm:221] opKeySet.find(mpscnnOp) != opKeySet.end(). 
```

The error says that what it means but is this something to do with the model?
Any hints on  [this](https://github.com/caffe2/caffe2/blob/417da38a988f2da7900a64dfa8cdfde635b0e8d6/caffe2/contrib/ios/mpscnn/mpscnn_graph.mm#L221)?

**Comment by tdp2110** [comment]
@KleinYuan : (note I'm not a caffe2 dev) basically what this is saying is that some operator in your net hasn't been registered. note that caffe2 mps can only handle a subset of all operators (which ones, I'm not sure, but you should be able to figure out from the code). you can try adding a log message in mpscnn_graph.mm to see which one in particular it's puking on. Good luck.

**Comment by tdp2110** [comment]
@KleinYuan I'm trying squeezenet and `tryConvertToMPSCNN`, and I got the same error as you. Debugging as I suggested, I see that `mpscnnOp` is "MPSCNNConv" and that no "MPSCNNXXX" op has ever been registered in `CPUOperatorRegistry()`. I don't know what this means, but I'll dig into it and if I get anywhere I'll let you know.

**Comment by tdp2110** [comment]
@KleinYuan so in my case, and I'm guessing in yours, the reason those MPS operators never get defined is that [mpscnn.mm](https://github.com/caffe2/caffe2/blob/master/caffe2/mobile/contrib/ios/mpscnn/mpscnn.mm) is never compiled. Why is it never compiled? `CAFFE2_USE_MPSCNN` is not defined. What happens if you try to define it? Well the includes fail, eg, `#include "caffe2/mobile/fb/aloha/generate_proposals_op.h"` -- there is no generate_proposals_op.h in the public github right now. I'm guessing this is not a user facing feature yet (if ever), which would explain the lack of documentation. This was noticed in #1162

**Comment by varunlakshmanan** [comment]
@jquave I'm getting the same error that @ankghost0912 is getting after following your tutorial. I'm building for a device, but am receiving a Mach-O Linker error. I've posted about my issue [here](https://github.com/caffe2/caffe2/issues/1666).

**Comment by Yangqing** [comment]
Should be fixed in #1690 now - feel free to reopen if it does not work.
